src/hmac.rs

```diff
@@ -175,15 +175,31 @@ impl Key {
         algorithm: Algorithm,
         rng: &dyn rand::SecureRandom,
     ) -> Result<Self, error::Unspecified> {
-        Self::construct(algorithm, |buf| rng.fill(buf))
+        Self::construct(algorithm, algorithm.0.output_len, |buf| rng.fill(buf))
     }
 
-    fn construct<F>(algorithm: Algorithm, fill: F) -> Result<Self, error::Unspecified>
+    /// Generate an HMAC singing key using a custom length for the given
+    /// digest algorithm.
+    ///
+    /// The key will be `length` bytes long.
+    pub fn generate_with_length(
+        algorithm: Algorithm,
+        length: usize,
+        rng: &dyn rand::SecureRandom,
+    ) -> Result<Self, error::Unspecified> {
+        Self::construct(algorithm, length, |buf| rng.fill(buf))
+    }
+
+    fn construct<F>(
+        algorithm: Algorithm,
+        length: usize,
+        fill: F,
+    ) -> Result<Self, error::Unspecified>
     where
         F: FnOnce(&mut [u8]) -> Result<(), error::Unspecified>,
     {
         let mut key_bytes = [0; digest::MAX_OUTPUT_LEN];
-        let key_bytes = &mut key_bytes[..algorithm.0.output_len];
```

Comment:
This doesn't work when the requested length is larger than `digest::MAX_OUTPUT_LEN`.

Please write a parameterized test that accepts an `hmac::Algorithm` and a length, and then tests your new function with that algorithm and length, for the following lengths:
1, 2, output_len - 1, output_len, output_len + 1, block_len - 1, block_len, block_len + 1, (block_len * 2) - 1, block_len * 2, (block_len * 2) + 1.

Reply:
Updated logic and added tests.

Reply:
You can't add a call to `to_vec` to `construct` because `construct` needs to be able to work even in environments that doesn't have a memory allocator. Instead you need to move the allocation to your new constructor and mark that constructor `#[cfg(feature = "alloc")]`.

Reply:
Ah nice catch! Moved new `construct` logic to `generate_with_length` behind the alloc feature :+1: